### DISCUSSION TITLE
Hows does a client communicate public Key to Keycloak that will be used to verify signed and encrypted JWTs?

### DISCUSSION THREAD

**Discussion by PrarthonaPaul** [keycloak/keycloak#22729] [Q&A] [answered]
Hi, 
I am trying to set up an application server so that when I send an authentication request to keycloak (or other openid providers), I can use `request` or `request-uri` parameters to send my query parameters as a jwt. These request parameters can optionally be signed and encrypted according to the[ OpenID docs](https://openid.net/specs/openid-connect-core-1_0.html#JWTRequests). See section 6.1 for how the claims can be converted to a jwt and signed using a keypair. 

However, in order to sign the jwt using JsonWebSignatures, I need to use a private key. But without communicating the public key to keycloak, the server cannot decode my Request Object. I was wondering where in the keyCloak admin console I can specify this public key as a client. 

Additionally, I am also trying to use the `org.keycloak.representations.idm.clientrepresentation` class to set up my client during run time. Is there a way I can specify the public key using this class? 
Thank you in advance for your help! 
If you have any follow-ups, please let me know!

**Comment by justin-tay** [comment, marked as answer]
To specify client keys in the console after you selected your client this should be configured in the `Keys` tab. You can either specify the JWKS URL of your client for Keycloak to fetch or set the JWKS directly. This tab only appears if `Capability config > Client authentication` is set to `On` which is on the `Settings` tab in the `Capability config` section below the `General Settings` and `Access settings` sections.

![keys](https://github.com/keycloak/keycloak/assets/49700559/a57f6d09-062d-4539-9bcc-5c8b7e0ec481)

Request object related configuration on the `Advanced` tab.

![keys2](https://github.com/keycloak/keycloak/assets/49700559/b3d65fac-be6c-4c93-a6f5-0539ba846746)

I think the class you are interested in should be the `OIDCAdvancedConfigWrapper`.

https://github.com/keycloak/keycloak/blob/b776746455fa920791d385a4a1302b7ce33e5b7c/services/src/main/java/org/keycloak/protocol/oidc/OIDCAdvancedConfigWrapper.java#L120-L154

https://github.com/keycloak/keycloak/blob/b776746455fa920791d385a4a1302b7ce33e5b7c/services/src/main/java/org/keycloak/protocol/oidc/OIDCAdvancedConfigWrapper.java#L80-L118

**Reply by PrarthonaPaul** [reply]
Thank you! This is very helpful! 
Just a quick follow-up, I noticed on the admin console there is a way to upload a keystore/PEM file directly. 
![image](https://github.com/keycloak/keycloak/assets/83880425/5b93b9ab-3c14-4a68-a2bc-69b9d3b21077)
Is there a way to do that programmatically and avoid using jwksUrl?

**Reply by justin-tay** [reply]
If you are refering to the admin console then you can always call the same Admin REST API that it is using.

The documentation is at https://www.keycloak.org/docs-api/22.0.1/rest-api/index.html.

`POST /admin/realms/{realm}/clients/{id}/certificates/{attr}/upload-certificate`

You can determine the payload to send by just performing the operations in the browser with the developer console (F12) open and look at the network tab.

In this case the payload is a multipart form-data.

```
------WebKitFormBoundaryWWmZKOPA75WLIWoO
Content-Disposition: form-data; name="keystoreFormat"

JKS
------WebKitFormBoundaryWWmZKOPA75WLIWoO
Content-Disposition: form-data; name="keyAlias"

keyalias
------WebKitFormBoundaryWWmZKOPA75WLIWoO
Content-Disposition: form-data; name="storePassword"

password
------WebKitFormBoundaryWWmZKOPA75WLIWoO
Content-Disposition: form-data; name="file"; filename="keystore.jks"
Content-Type: application/octet-stream
```

**Reply by PrarthonaPaul** [reply]
Thank you!

**Reply by PrarthonaPaul** [reply]
Hi there, I am a bit confused as to how to use the REST API. 
you mentioned i need to do a post request at `POST /admin/realms/{realm}/clients/{id}/certificates/{attr}/upload-certificate`. 
Do you know what attr represents? Is it one of the attributes listed here: 
```
"attributes": {
    "oidc.ciba.grant.enabled": "false",
    "client.secret.creation.time": "1693487816",
    "backchannel.logout.session.required": "true",
    "jwks.url": "http://localhost:8080/realms/WildFly/protocol/openid-connect/certs",
    "oauth2.device.authorization.grant.enabled": "false",
    "display.on.consent.screen": "false",
    "use.jwks.url": "false",
    "backchannel.logout.revoke.offline.tokens": "false",
    "jwt.credential.certificate": "MIIBGjCBwKADAgECAgYBik1GgRAwCgYIKoZIzj0EAwIwFDESMBAGA1UEAwwJbG9jYWxob3N0MB4XDTIzMDgzMDIwMjYyMFoXDTI1MDgzMDIwMjYyMFowFDESMBAGA1UEAwwJbG9jYWxob3N0MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEHGQZTnvZCC0dAzAV8bF7hLi2hDozJ2WXMmOWMAcZoC7n/GT3pl0WdB7XbxS7wXHyJaJPtJKI5EZcz/GzsPkIzTAKBggqhkjOPQQDAgNJADBGAiEAhLUMzNRGiSjkgu4HpnVNa3Pf+NXThp1zyM5FMEEa48wCIQCG14JGb9hNmj8D/tSllW/4YCCERfP4QR0kmohrnqtsAQ=="
  },
```

Additionally, once we make this POST request, what would be the jwksurl that we use to enter into the following function https://github.com/keycloak/keycloak/blob/b776746455fa920791d385a4a1302b7ce33e5b7c/services/src/main/java/org/keycloak/protocol/oidc/OIDCAdvancedConfigWrapper.java#L134-L136

Note that I am trying to configure multiple keys to use for signing and encrypting, which is why I need to use a jwksurl instead of importing it from a keystore file. 

Thank you for all your help so far! I really appreciate it!

**Reply by justin-tay** [reply]
You should just perform the operations using the admin console with F12 open and look at the network tab to see what the payload should be.

The JWKS url here is the JWKS url that your application exposes to Keycloak and is not something exposed by Keycloak. To set it you need to toggle `Use JWKS URL`.

![jwks](https://github.com/keycloak/keycloak/assets/49700559/4bfa199b-f0f6-42d3-b0e5-4783f67106da)
